flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/EpochManager.java

```diff
@@ -0,0 +1,202 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.asyncprocessing;
+
+import javax.annotation.Nullable;
+
+import java.util.LinkedList;
+
+/**
+ * Epoch manager segments inputs into distinct epochs, marked by the arrival of non-records(e.g.
+ * watermark, record attributes). Records are assigned to a unique epoch based on their arrival,
+ * records within an epoch are allowed to be parallelized, while the non-record of an epoch can only
+ * be executed when all records in this epoch have finished.
+ *
+ * <p>For more details please refer to FLIP-425.
+ */
+public class EpochManager {
+
+    /**
+     * This enum defines whether parallel execution between epochs is allowed. We should keep this
+     * internal and away from API module for now, until we could see the concrete need for {@link
+     * #PARALLEL_BETWEEN_EPOCH} from average users.
+     */
+    public enum ParallelMode {
+        /**
+         * Subsequent epochs must wait until the previous epoch is completed before they can start.
+         */
+        SERIAL_BETWEEN_EPOCH,
+        /**
+         * Subsequent epochs can begin execution even if the previous epoch has not yet completed.
+         * Usually performs better than {@link #SERIAL_BETWEEN_EPOCH}.
+         */
+        PARALLEL_BETWEEN_EPOCH
```

Comment:
The naming here is slightly different from the nouns used in FLIP-425. How about naming them as "strict order" and "out of order"?

Reply:
I think  "out of order" is a bit too broad, so I changed it to the current name.
As the [example](https://github.com/apache/flink/pull/24748#discussion_r1596141360 ) you mentioned, "out of order" can't describe the scenario that watermark1  executes after record 0-3.

 If there is a better way to describe "out of order", I am happy to change it.